### DISCUSSION TITLE
How can I get this project working in PySide6 Qt for Python and PyInstaller

### DISCUSSION THREAD

**Discussion by magbicaleman** [sparkle-project/Sparkle#2402] [Q&A]
I'm able to get the latest `Sparkle.framework` into the *.app bundle located in `Resources/FrameWorks/Sparkle.framework`.

In the info_plist file I have the following items set
```
..
 'SUFeedURL': 'https://localhost:3000/appcast.xml',
 'SUEnableAutomaticChecks': 'YES',
 'SUScheduledCheckInterval': '10',
 'CFBundleVersion': '0.0.1',
 'NSAppTransportSecurity': {
   'NSAllowsLocalNetworking': True,
  },
  ...
```

I'm serving the appcast.xml from `https://localhost:3000/appcast.xml` I'm using a generic xml
```
<?xml version="1.0" encoding="utf-8"?>
<rss xmlns:sparkle="http://www.andymatuschak.org/xml-namespaces/sparkle" version="2.0">
    <channel>
        <title>My Application Updates</title>
        <link>https://www.example.com/my-application/</link>
        <description>Updates for My Application</description>
        <language>en-us</language>
        <item>
            <title>Version 1.0.0</title>
            <sparkle:minimumSystemVersion>10.14</sparkle:minimumSystemVersion>
            <sparkle:releaseNotesLink>https://www.example.com/my-application/release-notes/1.0.0/</sparkle:releaseNotesLink>
            <pubDate>Mon, 01 Jan 2022 00:00:00 GMT</pubDate>
            <enclosure url="https://www.example.com/my-application/downloads/MyApplication-1.0.0.dmg"
                       sparkle:version="1.0.0"
                       sparkle:shortVersionString="1.0.0"
                       length="12345678"
                       type="application/octet-stream"/>
        </item>
        <item>
            <title>Version 1.1.0</title>
            <sparkle:minimumSystemVersion>10.14</sparkle:minimumSystemVersion>
            <sparkle:releaseNotesLink>https://www.example.com/my-application/release-notes/1.1.0/</sparkle:releaseNotesLink>
            <pubDate>Mon, 01 Feb 2022 00:00:00 GMT</pubDate>
            <enclosure url="https://www.example.com/my-application/downloads/MyApplication-1.1.0.dmg"
                       sparkle:version="1.1.0"
                       sparkle:shortVersionString="1.1.0"
                       length="12345678"
                       type="application/octet-stream"/>
        </item>
    </channel>
</rss>
```

In the python file I'm loading the framework

```
objc.loadBundle(
        "Sparkle", bundle_path='Resources/FrameWorks/Sparkle.framework', module_globals=globals())
        
 # later in the main window
 
 class MainWindow(QMainWindow):
   #code ...
   SPUStandardUpdaterController = objc.lookUpClass('SPUStandardUpdaterController')
   updater_controller = SPUStandardUpdaterController.alloc().init()
   updater_controller.checkForUpdates_(self)
```

The app starts correctly without an error, but the mock server, serving the app cast does not get a request, and the dialog to update does not show up... not sure what else I may need to do.

**Comment by magbicaleman** [comment]
Is there a way to debug Sparkle? I'm not seeing entries into my mac os x console app.

**Comment by zorgiepoo** [comment]
`SUScheduledCheckInterval` has a minimum bound of 3600 seconds (or 1 hour) by the way; specifying "10" seconds is not valid (it defaults to 1 day).

You're also not placing the framework in a valid location. It should be at `./Contents/Frameworks/Sparkle.framework` -- see https://developer.apple.com/documentation/bundleresources/placing_content_in_a_bundle (but not likely related to your current issue)

`SPUStandardUpdaterController.alloc().init()` is not valid. Check this class's other initializers. https://sparkle-project.org/documentation/api-reference/Classes/SPUStandardUpdaterController.html#/c:objc(cs)SPUStandardUpdaterController(im)init

`checkForUpdates` is only meant for user-initiated checks, although you could temporarily call it to see if Sparkle is working... but long term I would remove that line and let Sparkle check for updates on its automatic scheduled interval.

The appcast doesn't seem to currently contain any signing info.

Not familiar with python objc bindings, but make sure `updater_controller` is kept around / strongly referenced. The earliest this controller can be initialized is probably in the `MainWindow`'s initializer.

**Reply by magbicaleman** [reply]
@zorgiepoo thank you very much I appreciate the response.

> checkForUpdates is only meant for user-initiated checks, although you could temporarily call it to see if Sparkle is working... but long term I would remove that line and let Sparkle check for updates on its automatic scheduled interval.

Currently this is my goal, I just want to see if this is working, this is a new app and at the very least I'd like to see if this makes a request to my `https://localhost:3000/appcast.xml`, even if everything else doesn't work. 

> SUScheduledCheckInterval has a minimum bound of 3600 seconds (or 1 hour) by the way; specifying "10" seconds is not valid (it defaults to 1 day).

I removed `SUScheduledCheckInterval` from the  `info_plist`. Going to wait an hour and see if a request is made to the `SUFeedURL`, if that doesn't work I'll give the `Create an Updater in Qt` from https://sparkle-project.org/documentation/programmatic-setup/ a try. 

> You're also not placing the framework in a valid location. It should be at ./Contents/Frameworks/Sparkle.framework -- see https://developer.apple.com/documentation/bundleresources/placing_content_in_a_bundle (but not likely related to your current issue)

I'm looking into this, opened a discussion in the pyinstaller repo.
- https://github.com/orgs/pyinstaller/discussions/7755

> SPUStandardUpdaterController.alloc().init() is not valid. Check this class's other initializers. https://sparkle-project.org/documentation/api-reference/Classes/SPUStandardUpdaterController.html#/c:objc(cs)SPUStandardUpdaterController(im)init

I tried `SPUStandardUpdaterController.alloc().startUpdater()` and it starts my app then crashes it with the report a crash dialog, not sure if this means if my configuration is bad or if doing that just causes a crash. 

> The appcast doesn't seem to currently contain any signing info.

I'm not seeing any requests to my `https://localhost:3000/appcast.xml` at all, so this wouldn't matter yet? I also added `'SUPublicEDKey': 'pfIShU4dEXqPd5ObYNfDBiQWcXozk7estwzTnF9BamQ=',` to my `info_plist` from the documentation page, just to see if that would help make a request to `SUFeedURL`, so that it had something to potentially compare.

**Reply by zorgiepoo** [reply]
> I removed SUScheduledCheckInterval from the  info_plist. Going to wait an hour and see if a request is made to the SUFeedURL, if that doesn't work I'll give the Create an Updater in Qt from https://sparkle-project.org/documentation/programmatic-setup/ a try.

Check the `defaults delete` command in https://sparkle-project.org/documentation#6-test-sparkle-out to delete the last updated check time, so you won't have to wait (it defaults to 1 day, not 1 hour).

> I tried SPUStandardUpdaterController.alloc().startUpdater() and it starts my app then crashes it with the report a crash dialog, not sure if this means if my configuration is bad or if doing that just causes a crash.

You still need to use an init method [like this one](https://sparkle-project.org/documentation/api-reference/Classes/SPUStandardUpdaterController.html#/c:objc(cs)SPUStandardUpdaterController(im)initWithUpdaterDelegate:userDriverDelegate:), just not `init()` which is marked unavailable. It might be `SPUStandardUpdaterController.alloc().initWithUpdaterDelegate_userDriverDelegate_(None, None)` if I got the python-objc syntax correctly.

Let's see if that gets you further. Further errors, once the updater is started, should be found in Console logs for the app.
